stl/inc/span

```diff
@@ -0,0 +1,732 @@
+        _Myptr = _It;
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+private:
+#if _ITERATOR_DEBUG_LEVEL >= 1
+    pointer _Mydata = nullptr;
+    size_t _Mysize  = 0;
+    size_t _Myoff   = 0;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+    pointer _Myptr = nullptr;
+#endif // _ITERATOR_DEBUG_LEVEL
+};
+
+// FUNCTION TEMPLATE _Is_span_v
+template <typename _Ty, size_t _Extent>
+class span;
+
+template <class>
+inline constexpr bool _Is_span_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_span_v<span<_Ty, _Extent>> = true;
+
+// FUNCTION TEMPLATE _Is_std_array_v
+template <class>
+inline constexpr bool _Is_std_array_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_std_array_v<array<_Ty, _Extent>> = true;
+
+// FUNCTION TEMPLATE _Is_span_convertible_range_v
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_convertible_range_v =
+    is_convertible_v<remove_pointer_t<decltype(data(declval<_Cont&>()))> (*)[], _Ty (*)[]>;
+
+// FUNCTION TEMPLATE _Has_container_interface
+template <class, class = void>
+inline constexpr bool _Has_container_interface_v = false;
+
+template <class _Cont>
+inline constexpr bool
+    _Has_container_interface_v<_Cont, void_t<decltype(data(declval<_Cont>())), decltype(size(declval<_Cont>()))>> =
+        true;
+
+// FUNCTION TEMPLATE _Is_span_compatible_container_v
+// clang-format off
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_compatible_container_v =
+    !_Is_span_v<_Cont> &&
+	!_Is_std_array_v<_Cont> &&
+	!is_array_v<_Cont> &&
+	_Is_span_convertible_range_v<_Cont, _Ty> &&
+	_Has_container_interface_v<_Cont>;
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span {
+public:
+    using element_type                 = _Ty;
+    using value_type                   = remove_cv_t<_Ty>;
+    using index_type                   = size_t;
+    using difference_type              = ptrdiff_t;
+    using pointer                      = _Ty*;
+    using const_pointer                = const _Ty*;
+    using reference                    = _Ty&;
+    using const_reference              = const _Ty&;
+    using iterator                     = _Span_iterator<pointer>;
+    using const_iterator               = _Span_iterator<const_pointer>;
+    using reverse_iterator             = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator       = _STD reverse_iterator<const_iterator>;
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    constexpr span() noexcept = default;
+
+    constexpr span(pointer _Ptr, index_type _Count) noexcept : _Mydata(_Count, _Ptr) {}
+
+    constexpr span(pointer _First, pointer _Last) noexcept : _Mydata(static_cast<index_type>(_Last - _First), _First) {}
+
+    template <size_t _Size>

```

Comment:
@CaseyCarter already mentioned this. However my reasoning was the following:
The first constraint is:
```
extent == dynamic_­extent || N == extent is true
```

This constraint is already handeled by `_Span_extent`

The other constraint is trivially fulfilled the way the arguments to the constructors are defined using `element_type` and `value_type`.
```
remove_­pointer_­t<decltype(data(arr))>(*)[] is convertible to ElementType(*)[]
```

Maybe I am missing something but is there any posibility that this requirement is not met?

Reply:
Looking at [libcxx](https://github.com/llvm-mirror/libcxx/blob/fb944628a6cf75b707836062e73e8aebc54c1fbd/include/span#L223) there are also no additional constraints checked

Reply:
Constraints mean SFINAE. Whatever `_Span_extent` does, it will not cause this constructor signature to vanish. The Constraints say that it should vanish, which will cause `is_constructible` to report correct answers.

Reply:
Thanks a lot for the clarification. I am still learning the minutiae of the standard wording, which in hindsight this is all about.

As a sideline should the standard drop the unecessary requirement?

Reply:
When I looked at this, I thought for at least one constructor, the convertibility requirement was handling const qualification.

Reply:
This was resolved by adding constraints for the `const array` constructor